### PR TITLE
dynamic carousel

### PR DESCRIPTION
**Pull request by Rogermax**
## Description

Add to current Carousel component a DynamicCarousel as Shadcn Carousel (neither https://www.embla-carousel.com/api/) allow us to have the desired behaviour. We made this alternative.

https://github.com/user-attachments/assets/3face02f-e607-458e-a1a3-c1d9fce40c26

### Figma Link

[Link to Figma Design](https://www.figma.com/design/sjXPhNmfJI3OnVpWmfn6Lc/Accounts-Payable?node-id=4351-32519&t=SibA7rf6DmeBlYAm-4)

## Type of Change

- [x] Improve experimental component

### CODE REVIEW

File: lib/experimental/Navigation/DynamicCarousel/index.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+  const scrollToNextItem = (direction: "left" | "right") => {
+    if (scrollContainerRef.current) {
+      const container = scrollContainerRef.current
+      const { scrollLeft } = container
+
+      if (direction === "right") {
+        const nextItem = Array.from(container.children).find((child) => {
+          const childLeft = (child as HTMLElement).offsetLeft
+          return childLeft - 28 > scrollLeft
```
Comment (nlopin) [inline lib/experimental/Navigation/DynamicCarousel/index.tsx:20]: 28 is all around, make sense to have it as a variable

**Comment by nlopin** [comment]
Please, add description to the component. In one month we won't remember why it is here. Ideally, it should be a markdown document explaining the use case and best practices. Take `DataList` as example (lib/experimental/Lists/DataList/index.mdx)

**Comment by Rogermax** [comment]
@nlopin I removed from the components list (is a sub component of Carousel) so I think it's not necessary to document the inner Carousel no?

**Comment by nlopin** [comment]
> @nlopin I removed from the components list (is a sub component of Carousel) so I think it's not necessary to document the inner Carousel no?

true, my bad

**Review by nlopin — APPROVED** [review]

File: lib/ui/carousel.tsx [hunk 2]
```diff
@@ -251,7 +251,7 @@ const CarouselNext = React.forwardRef<
       {...props}
     >
       <Icon size="sm" icon={ArrowRight} />
-      <span className="sr-only">Mext</span>
```
Comment (josepjaume) [inline lib/ui/carousel.tsx:254]: Lol

File: lib/experimental/Navigation/Carousel/index.tsx [hunk 3]
```diff
@@ -60,6 +61,10 @@ export const Carousel = ({
     return showPeek ? (`peek${value || 1}` as PeekVariant) : value || 1
   }
 
+  if (!columns) {
+    return <DynamicCarousel>{children}</DynamicCarousel>
```
Comment (josepjaume) [inline lib/experimental/Navigation/Carousel/index.tsx:65]: Shouldn't we be passing the same props of the carousel here? So things like `onMouseEnter` still work and all that. Probably `opts` as well, and everything else.

File: lib/experimental/Navigation/Carousel/index.stories.tsx [hunk 4]
```diff
@@ -74,10 +72,27 @@ export const AutoScroll: Story = {
...
+      <div className="w-full p-6">
+        <Story />
+      </div>
+    ),
+  ],
+  args: {
+    showArrows: false,
+    children: SLIDES,
+    showPeek: true,
+    columns: { default: 1 },
```
Comment (josepjaume) [inline lib/experimental/Navigation/Carousel/index.stories.tsx:91]: I've said that in another PR but I'm surprised to see that many choices here. Unless I'm missing something, this component should just show or hide arrows depending on the amount of slides, no?

File: lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            "absolute",
+            "h-full",
+            "top-0",
+            "bg-gradient-to-r from-transparent from-0% to-f1-background to-100%"
+          )}
+          style={{ right: `-${SPACE_FOR_WIDGET_SHADOW}px` }}
+        ></div>
+      )}
+
+      {/* Botón Izquierdo dinámico */}
```
Comment (josepjaume) [inline lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx:142]: What's this weird language?

File: lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            "-left-3 top-1/2 -translate-y-1/2"
+          )}
+          onClick={() => scrollToNextItem("left")}
+        >
+          <Icon size="sm" icon={ArrowLeft} />
+          <span className="sr-only">Previous</span>
+        </Button>
+      )}
+
+      {/* Botón Derecho dinámico */}
```
Comment (josepjaume) [inline lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx:159]: Same!

File: lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+              <div
+                className="flex-shrink-0"
+                style={{ scrollSnapAlign: "start" }}
+              >
+                {children}
+              </div>
+            )}
+      </div>
+
+      {/* Gradientes dinámicos para los bordes */}
```
Comment (josepjaume) [inline lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx:116]: 👁️

File: lib/experimental/Navigation/Carousel/index.tsx (hunk 3 above)
Comment (Rogermax) [reply lib/experimental/Navigation/Carousel/index.tsx:65]: mmm.. Carousel don't have that props.. is just to make the plugin of autoplay works... I can try to implement and autoplay to our new component, but not in that PR, right?

File: lib/experimental/Navigation/Carousel/index.stories.tsx (hunk 4 above)
Comment (Rogermax) [reply lib/experimental/Navigation/Carousel/index.stories.tsx:91]: for me is ok, we just need to decide :) I can change it

File: lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx (hunk 6 above)
Comment (Rogermax) [reply lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx:159]: fixing here: https://github.com/factorialco/factorial-one/pull/979

File: lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx (hunk 7 above)
Comment (Rogermax) [reply lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx:116]: fixing here: https://github.com/factorialco/factorial-one/pull/979

**Pull request merged**